### PR TITLE
Improve MP_HMAC 

### PR DESCRIPTION
**Pull request by veselinrak**
Text updated to clarify key generation for different suboptions and to correct/clarify several other issues.

### CODE REVIEW

File: draft-ietf-tsvwg-multipath-dccp.mkd [hunk 1]
```diff
@@ -818,26 +818,26 @@ chosen in MP_SEQ are considered sufficiently large.
...
 
 The MP_HMAC suboption is used to provide authentication for the MP_JOIN,
-MP_ADDADDR, and MP_REMOVEADDR suboptions. The HMAC code is generated according
+MP_ADDADDR, and MP_REMOVEADDR suboptions. For this specification of MP-DCCP, the HMAC code is generated according
 to {{RFC2104}} in combination with the SHA256 hash algorithm described in
 {{RFC6234}}, with the output truncated to the leftmost 160 bits (20 bytes).
 
 The "Key" used for the HMAC computation is the derived key (d-key)
-described in {{MP_KEY}}, while the HMAC "Message" is a concatenation of
+described in {{MP_KEY}}, while the HMAC "Message" for MP_JOIN, MP_ADDADR and MP-REMOVEADDR is a concatenation of:
```
Comment (markusa) [inline draft-ietf-tsvwg-multipath-dccp.mkd:826]:
```suggestion
described in {{MP_KEY}}, while the HMAC "Message" for MP_JOIN, MP_ADDADDR and MP_REMOVEADDR is a concatenation of:
```

**Review by markusa — APPROVED** [review]
Based on what we discussed in the authors' meeting, it seems to be complete.

File: draft-ietf-tsvwg-multipath-dccp.mkd [hunk 2]
```diff
@@ -818,26 +818,26 @@ chosen in MP_SEQ are considered sufficiently large.
...
    and if defined port, otherwise two octets of value 0.
    
-   * MP_REMOVEADDR: Solely the Address ID.
+   * for MP_REMOVEADDR: Solely the Address ID.
 
 MP_JOIN, MP_ADDADDR and MP_REMOVEADDR can co-exist or be used multiple times
-within a single DCCP packet. All these multipath options include an individual
-MP_HASH option. This ensures the MP_HASH is correctly associated.
+within a single DCCP packet. All these multipath options require an individual
+MP_HMAC option. This ensures the MP_HMAC is correctly associated.
```
Comment (abrunstrom) [inline draft-ietf-tsvwg-multipath-dccp.mkd:838]:
```suggestion
MP_HMAC option. This ensures that the MP_HMAC is correctly associated.
```

File: draft-ietf-tsvwg-multipath-dccp.mkd [hunk 3]
```diff
@@ -818,26 +818,26 @@ chosen in MP_SEQ are considered sufficiently large.
...
+MP_ADDADDR, and MP_REMOVEADDR suboptions. For this specification of MP-DCCP, the HMAC code is generated according
 to {{RFC2104}} in combination with the SHA256 hash algorithm described in
 {{RFC6234}}, with the output truncated to the leftmost 160 bits (20 bytes).
 
 The "Key" used for the HMAC computation is the derived key (d-key)
-described in {{MP_KEY}}, while the HMAC "Message" is a concatenation of
+described in {{MP_KEY}}, while the HMAC "Message" for MP_JOIN, MP_ADDADDR and MP_REMOVEADDR is a concatenation of:
 
-   * MP_JOIN: The CI and nonce of the MP_JOIN for which authentication
+   * for MP_JOIN: The CI and nonce of the MP_JOIN for which authentication
```
Comment (abrunstrom) [inline draft-ietf-tsvwg-multipath-dccp.mkd:828]: When we discussed this part in the telco last week, we decided to also add the psudocode style description. Not sure if it was intentionally left out or forgotten. Also not sure if it is really needed, so just putting a reminder of the discussion :)

**Review by abrunstrom — COMMENTED** [review]
Added a comment and a nit.

File: draft-ietf-tsvwg-multipath-dccp.mkd (hunk 3 above)
Comment (markusa) [reply draft-ietf-tsvwg-multipath-dccp.mkd:828]: That is correct. For example the one for the MP_JOIN can be found in section 4.3:
`MP_HMAC(B) = HMAC-SHA256(Key=d-key(B), Msg=RB+RA)`

Comment (NathalieRM) [reply draft-ietf-tsvwg-multipath-dccp.mkd:828]: This example "MP_HMAC(B) = HMAC-SHA256(Key=d-key(B), Msg=RB+RA)" and the description of the HMAC calculation for an MP_JOIN in section 4.3, don't match with the description written here. Here it is stated that  Msg = CI(A or B)+R(A or B),

Comment (markusa) [reply draft-ietf-tsvwg-multipath-dccp.mkd]: I think CI is not needed and I removed it from the calculation. I also added HMAC-SHA265 calculation examples for MP_ADDADDR and MP_REMOVEADDR in https://github.com/tsvwg/ietf-multipath-dccp/pull/280/commits/23f89ad63eaf38c65e3d5afff15a8c55ddde75eb
@abrunstrom @NathalieRM

**Comment by markusa** [comment]
In the course of discussing the HMAC-SHA256 calculation, I noticed that the HMAC usage for `MP_JOIN` differs from that of `MP_ADDADDR` and `MP_REMOVEADDR` in that no HMAC is assigned for the first MP_JOIN. There are good reasons for this, as the nonces have to be exchanged first. I don't think we contradict this with our text at the moment, but I suspect that because we don't say this explicitly in the `MP_HMAC` or `MP_JOIN` section, the reader might think that an `MP_HMAC` must already be sent with the first MP_JOIN.
The first place where it becomes clear is in the handshake example in figure 21.
@abrunstrom @NathalieRM @veselinrak

**Comment by markusa** [comment]
> Looking at Fig. 21 I am now confused. There it looks like the MP_HMAC is sent alone, but from the MP_HMAC section and our earlier discussions I thought it always followed another sub-option? If sent alone it should be ignored.

I think the definition in Fig. 21 makes sense as it also follows the MPTCP HMAC authentication scheme for MP_JOIN. First the nonces RA and RB must be available before MP_HMAC can be used. The question is, what texts or passages need to be changed to eliminate your confusion?

Is it for example the sentence: ` Therefore, a MP_HMAC MUST directly follow its associated multipath option.`?

**Comment by NathalieRM** [comment]
Hi @Markus, 

I think this text `MP_HMAC MUST be set when using MP_JOIN in the second and third step of the handshake of a subsequent subflow as illustrated in {{handshaking}}.` is not completely accurate. As Anna stated, in the third step we are not using any MP_JOIN, that is just a DCCP-Ack packet with no MP_OPTION.  I would propose something like `MP_HMAC MUST be sent when using MP_JOIN within a DCCP-Response Packet (See Section 4.2.6 for details).`  instead. 

Apart from that, I agree with you with the fact that the Handshaking procedure makes sense as it is, and we just need to clarify that the MP_HMAC is not always sent along with other suboptions. In addition to the text you already outlined, I would say it is also necessary to change this text `The MP_HMAC suboption is used to provide authentication for the MP_JOIN, MP_ADDADDR, and MP_REMOVEADDR suboptions`, my suggestion would be ` The MP_HMAC suboption is used to provide authentication for the  MP_ADDADDR, and MP_REMOVEADDR suboptions. In addition, it provides authentication for subflows joinning an existing MP_DCCP connection, as described in the second and third step of the handshake of a subsequent subflow in {{handshaking}}.`

**Comment by abrunstrom** [comment]
We also just added text saying 

> In the event that an MP_HMAC cannot be associated with a suboption, this MP_HMAC MUST be ignored.

This addition is no longer correct as stated and needs to be more fine grained. 

It is also a bit confusing what A is in the handshaking figure. If the text describing the HMAC is correct, A is not the same A as described in the MP_HMAC section. Do we need to also include the A from the handshake figure in the MP_HMAC section?

**Comment by markusa** [comment]
@NathalieRM I adopted your suggestion from https://github.com/tsvwg/ietf-multipath-dccp/pull/280#issuecomment-1863616121 in https://github.com/tsvwg/ietf-multipath-dccp/pull/280/commits/c02e45998beaf48dd5307236faf2af25a743b796

@abrunstrom I modified the sentence you mentioned in https://github.com/tsvwg/ietf-multipath-dccp/pull/280#issuecomment-1865272204 to cover the aspect of the single MP_HMAC sent in response to a MP_JOIN in https://github.com/tsvwg/ietf-multipath-dccp/pull/280/commits/ef9e5e9f933069d563598fc888af3401d0144fd2.

@abrunstrom Can you please explain what is confusing about "A" in the handshaking figure? We already introduce the terms "Host A" and "B" in Figure 2, and I thought we would use them consistently from then on.

File: draft-ietf-tsvwg-multipath-dccp.mkd [hunk 4]
```diff
@@ -849,7 +864,8 @@ cannot be validated by a receiving host because the HMAC validation fails, the s
 on which suboption was being verified. If the suboption to be authenticated was either
 MP_ADDADDR or MP_REMOVEADDR, the receiving host MUST silently ignore it (see {{MP_ADDADDR}} and {{MP_REMOVEADDR}}). 
 If the suboption to be authenticated was MP_JOIN, the subflow MUST be closed (see {{fallback}}).
-In the event that an MP_HMAC cannot be associated with a suboption, this MP_HMAC MUST be ignored.
+In the event that an MP_HMAC cannot be associated with a suboption, unless it is an MP_HMAC sent
+in DCCP-Ack in response to received MP_JOIN, this MP_HMAC MUST be ignored.
```
Comment (NathalieRM) [inline draft-ietf-tsvwg-multipath-dccp.mkd:868]:
```suggestion
in DCCP-Ack in response to a DCCP-Response packet containing an MP_JOIN option, this MP_HMAC MUST be ignored.
```

**Pull request merged**